### ISSUE TITLE
Linux: create/delete coalescing into modification events has broken inotify watching

### ISSUE THREAD

**Issue by Kynde** [parcel-bundler/watcher#117]
I ran into some problems with file watching in parcel on linux (recent Fedora). Parcel 2.8.0 with watcher 2.0.7.

Steps to reproduce the problem:

1. parcel serve running
2. make a change to a .jsx file that's easy to detect
3. change is detected nicely and page reloads
4. git checkout <aforementioned .jsx file>
5. parcel loses all tracking to that file

If caching is used, the sad state remains. With --no-cache it works again after a restart. Similar problems can be done with git stash and git pop. 

`git checkout` deletes the dirty file when it's checking it out again. With inotifywait it's easy to see that the file gets deleted and then replaced with a new one.

This likely makes the inotify based system lose track of the file. Either the parent directory changes of a file are not being watched or they are handled incorrectly.

I first ran into this problem with my `vim`, but that I could configure so that it doesn't delete the file, but `git` is a bit much.

Based on a coworker this is not a problem on a Mac.

**Comment by mischnic** [comment]
Do you know if this is was always broken or if it's a regression from Parcel 2.8 and watcher 2.0.6/2.0.7.

**Comment by Kynde** [comment]
I have no idea. We just migrated to parcel.

I suppose I could try an older version if you want. Anything in mind?

**Comment by Kynde** [comment]
Ok, between 2.0.3 and 2.0.4 someone implemented coalescing of delete and create into single update event.

More on that in issue #72 

I think this is the culprit. While it's ok to do that in principle, and on OSes where monitoring doesn't break because of DELETE/CREATE, that's not kosher on linux and inotify. I don't see any code that would account for the need to remonitor the file.

The inotify watching dies at DELETE and unless it's re-established upon CREATE, all future changes just get lost.

That's just a hypothesis at this point. I'll try the version 2.0.3 and see how that behaves.

**Comment by Kynde** [comment]
Indeed, 2.0.3 works.

So the offending commit is this: 4b749ab2d2f63a56cf98c3374849e2815ebb1994

**Comment by bpasero** [comment]
The intent of my change was provide a coalesced event to the outside listener and I was not aware it could break something internally. It would probably be safer to do coalescing at one point before emitting the event, but keep the events as they are for internal processing, such as having to re-listen when a file is deleted on Linux.

**Comment by bpasero** [comment]
Actually the original issue in https://github.com/parcel-bundler/watcher/issues/72 is that the watcher was emitting a create-event and not a delete+create. So there was some coalescing happening but imho the wrong one. If instead delete+create would be emitted in one go, applications like VS Code could build the right coalescing on top.

**Comment by veldsla** [comment]
I think this has bitten me too.

`parcel watch` failed to correctly rebuild after changes (.js entry points). Although some event is triggered, no new bundles are created. Console logs only bundling events (always 2):
```
[parcelwatch] Bundling...
[parcelwatch] ✨ Built in 302ms
[parcelwatch] ✨ Built in 302ms
[parcelwatch] Bundling...
[parcelwatch] ✨ Built in 36ms
[parcelwatch] ✨ Built in 36ms
```

When if force the watcher subdependency to 2.0.3 I get correct rebuilds and the following output:
```
[parcelwatch] Building...
[parcelwatch] Bundling...
[parcelwatch] Packaging & Optimizing...
[parcelwatch] ✨ Built in 121ms
[parcelwatch] ✨ Built in 121ms
```

This is linux on a nvme ssd disk (suspected the watcher dependency when a coworker had no issues over NFS and found this issue)

**Comment by colatkinson** [comment]
I've also run into this, also on an NVMe drive. The "fix" of downgrading to 2.0.3 seems to work for me as well.

Is this the currently recommended workaround, or are there any config settings, etc. that can be tweaked to get the desired behavior?

**Comment by Kynde** [comment]
We now upgraded to 2.11.0 and I cannot reproduce the original problem anymore, so I'm calling this fixed.
